### PR TITLE
Panorama alignment with a reference camera

### PR DESCRIPTION
**Pull request by servantftechnicolor**
User can align the panorama horizontally with a target view. Need to select auto_from_cameras_x_axis in sfmTransform.

### CODE REVIEW

File: src/software/utils/main_sfmTransform.cpp [hunk 1]
```diff
@@ -184,6 +184,36 @@ static void parseManualTransform(const std::string& manualTransform, double& S,
 
 } // namespace
 
+IndexT getReferenceViewIndex(const sfmData::SfMData & sfmData, const std::string & transform)
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:187]: getReferenceViewId

File: src/software/utils/main_sfmTransform.cpp [hunk 2]
```diff
@@ -184,6 +184,36 @@ static void parseManualTransform(const std::string& manualTransform, double& S,
...
+{
+    IndexT refPoseId;
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:189]: refViewId

File: src/software/utils/main_sfmTransform.cpp [hunk 3]
```diff
@@ -184,6 +184,36 @@ static void parseManualTransform(const std::string& manualTransform, double& S,
...
+        refPoseId = UndefinedIndexT;
+    }
+
+    if (refPoseId == UndefinedIndexT)
+    {
+        // Sort views per timestamps
+        std::vector<std::pair<int64_t, IndexT>> sorted_views;
+        for (auto v : sfmData.getViews()) {
+            int64_t t = v.second->getMetadataDateTimestamp();
+            sorted_views.push_back(std::make_pair(t, v.second->getPoseId()));
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:205]: sorted_views.push_back(std::make_pair(t, v.second->getViewId()));

File: src/software/utils/main_sfmTransform.cpp [hunk 4]
```diff
@@ -184,6 +184,36 @@ static void parseManualTransform(const std::string& manualTransform, double& S,
 
 } // namespace
 
+IndexT getReferenceViewIndex(const sfmData::SfMData & sfmData, const std::string & transform)
+{
+    IndexT refPoseId;
+    try
+    {
+        refPoseId = sfm::getViewIdFromExpression(sfmData, transform);
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:192]: refViewId = sfm::getViewIdFromExpression(sfmData, transform);

File: src/software/utils/main_sfmTransform.cpp [hunk 5]
```diff
@@ -269,6 +299,10 @@ int aliceVision_main(int argc, char **argv)
     return EXIT_FAILURE;
   }
 
+  IndexT refPoseId = getReferenceViewIndex(sfmData, transform);
+
+  Eigen::Matrix3d ref_R_world = sfmData.getAbsolutePose(refPoseId).getTransform().rotation();
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:304]: Eigen::Matrix3d ref_R_world = sfmData.getPose(sfmData.getView(refViewId)).getTransform().rotation();

File: src/software/utils/main_sfmTransform.cpp [hunk 6]
```diff
@@ -306,8 +340,25 @@ int aliceVision_main(int argc, char **argv)
...
+        Eigen::Vector3d alignmentVector = updatedWorld_R_refcam * Eigen::Vector3d::UnitZ();
+
+        Eigen::Matrix3d zeroX_R_world;
+        sfm::getRotationNullifyX(zeroX_R_world, alignmentVector);
+
+        //Cumulate both transformations
+        //(zeroX_R_world * updatedWorld_R_refcam).transpose()
+        //(zeroX_R_world * (ref_R_world * R.transpose()).transpose()).transpose()
+        // ref_R_world * R.transpose() * zeroX_R_world.transpose()
+        R = zeroX_R_world * R;
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:359]: to move to a dedicated function in the library and not in the main here.

**Review by fabiencastan — CHANGES_REQUESTED** [review]
2 fixes needed

File: src/software/utils/main_sfmTransform.cpp [hunk 7]
```diff
@@ -184,6 +184,36 @@ static void parseManualTransform(const std::string& manualTransform, double& S,
...
+        refViewId = UndefinedIndexT;
+    }
+
+    if (refViewId == UndefinedIndexT)
+    {
+        // Sort views per timestamps
+        std::vector<std::pair<int64_t, IndexT>> sorted_views;
+        for (auto v : sfmData.getViews()) {
+            int64_t t = v.second->getMetadataDateTimestamp();
+            sorted_views.push_back(std::make_pair(t, v.second->getPoseId()));
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransform.cpp:205]: still getViewId ?

**Review by cbentejac — APPROVED** [review]
Tested on Linux with https://github.com/alicevision/Meshroom/pull/1876 on both pinhole and fisheye panoramas, with and without HDR.

For all the used datasets, the PR worked as intended (a rebase was needed, though).

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**